[doc] Document the test dependency feature (#1084)
eth-cscs/reframe: docs/dependencies.rst

=====================================
How Test Dependencies Work In ReFrame
=====================================

Before going into details on how ReFrame treats test dependencies, it is important to understand how tests are actually treated and executed by the runtime.
Normally, a ReFrame test will be tried for different programming environments and different partitions within the same ReFrame run.
These are defined in the test's :func:`__init__` method, but it is not this original object that is being executed by the regression test pipeline.
The following figure explains in more detail the process:

.. figure:: _static/img/reframe-test-cases.svg
  :align: center
  :alt: How ReFrame loads and schedules tests for execution.

When ReFrame loads a test from the disk it unconditionally constructs it executing its :func:`__init__` method.

> The practical implication of this is that your test will be instantiated even if it will not be run on the current system.

Reviewer: ```suggestion
The practical implication of this is that your test will be instantiated even if it is not going to run on the current system.
```

Author: I will use "even if it will not run on the current system," instead.

After all the tests are loaded, they are filtered based on the current system and any other criteria (such as programming environment, test attributes etc.) specified by the user (see `Filtering of Regression Tests <running.html#filtering-of-regression-tests>`__ for more details).
After the tests are filtered, ReFrame creates the actual `test cases` to be run. A test case is essentially a tuple consisting of the test, the system partition and the programming environment to try.
The test that goes into a test case is essentially a `clone` of the original test that was instantiated upon loading.
This ensures that the test case's state is not shared and may not be reused in any case.
Finally, the generated test cases are passed to a `runner` that is responsible for scheduling them for execution based on the selected execution policy.

> Dependencies in ReFrame are defined at the test level using the :func:`depends_on` function, but are projected to the test cases space.

Reviewer: ```suggestion
Dependencies in ReFrame are defined at the test level using the :func:`depends_on` function, but are projected to the test cases' space.
```

Author: I will skip that, too, because it's not possessive. The space does not belong to the test cases, but rather the test cases form a space.

We will see the rules of that projection in a while.

> The dependency graph construction and the subsequent dependency analysis happens also at the level of the test cases.

Reviewer 2: ```suggestion
The dependency graph construction and the subsequent dependency analysis happen also at the level of the test cases.
```

Author: This is plural, because I have two subjects: "The dependency graph construction and the subsequent dependency analysis."

Let's assume that test :class:`T1` depends in :class:`T0`.

> This can be expressed inside :class:`T1` using the :func:`depends_on` function:

Reviewer 2: ```suggestion
This can be expressed inside :class:`T1` using the :func:`depends_on` method:
```

.. code:: python

   @rfm.simple_test
   class T1(rfm.RegressionTest):
       def __init__(self):
           ...
           self.depends_on('T0')

Conceptually, this dependency can be viewed at the test level as follows:

.. figure:: _static/img/test-deps.svg
  :align: center
  :alt: Simple test dependency presented conceptually.

For most of the cases, this is sufficient to reason about test dependencies.

> In reality, as mentioned above, dependencies are handled at the level of test cases.

Reviewer 2: ```suggestion
In reality, as mentioned above, dependencies are handled at the test case level.
```

Author: Here, I might leave what I had already.

Test cases on different partitions are always independent.

> If not specified differently, test cases with using programming environments are also independent.

Reviewer 2: ```suggestion
If not specified differently, test cases using programming environments are also independent.
```

.. figure:: _static/img/test-deps-by-env.svg
  :align: center
  :alt: Test case dependencies by environment (default).

This means that test cases of :class:`T1` may start executing before all test cases of :class:`T0` have finished.
You can impose a stricter dependency between tests, such that :class:`T1` does not start execution unless all test cases of :class:`T0` have finished.
You can achieve this as follows:

.. code:: python

   @rfm.simple_test
   class T1(rfm.RegressionTest):
       def __init__(self):
           ...
           self.depends_on('T0', how=rfm.DEPEND_FULLY)

This will create the following test case graph:

.. figure:: _static/img/test-deps-fully.svg
  :align: center
  :alt: Fully dependent test cases.

You may also create arbitrary dependencies between the test cases of different tests, like in the following example, where the dependencies cannot be represented in any of the other two ways:

.. figure:: _static/img/test-deps-exact.svg
  :align: center
  :alt: Arbitrary test cases dependencies

> These dependencies be achieved as follows:

Reviewer 2: ```suggestion
These dependencies can be achieved as follows:
```

.. code:: python

   @rfm.simple_test
   class T1(rfm.RegressionTest):
       def __init__(self):
           ...
           self.depends_on('T0', how=rfm.DEPEND_EXACT,
                           subdeps={'E0': ['E0', 'E1'], 'E1': ['E1']})

The ``subdeps`` argument defines the sub-dependencies between the test cases of :class:`T1` and :class:`T0` using an adjacency list representation.

Cyclic dependencies
-------------------

Obviously, cyclic dependencies between test cases are not allowed.
Cyclic dependencies between tests are not allowed either, even if the test case dependency graph is acyclic.
For example, the following dependency set up is invalid:

.. figure:: _static/img/test-deps-cycle.svg
  :align: center
  :alt: Any cyclic dependencies between tests are not allowed, even if the underlying test case dependencies are not forming a cycle.

> The test case dependencies here are clearly not forming a cycle, but the edge from ``(T0, E0)`` to ``(T1, E1)`` introduces a dependency from ``T0`` to ``T1`` forming a cycle at the test level.

Reviewer 2: ```suggestion
The test case dependencies here are clearly don't form a cycle, but the edge from ``(T0, E0)`` to ``(T1, E1)`` introduces a dependency from ``T0`` to ``T1`` forming a cycle at the test level.
```

Reviewer: If you change this as Theo suggests, remove the `are` from `here are clearly don't form a cycle`.

Author: You probably mean to remove "are" here, too.
